### ISSUE TITLE
Implement custom blocks for end users

### ISSUE THREAD

**Issue by johno** [blocks/blocks#14]
For certain use cases it'd be great to provide a way for users to mix in a handful of their own components to the Blocks editor. We want something that essentially "Just Works" for most components, and complex needs for users can be addressed using a custom plugin.

To start I think we should only focus on block-level custom components to keep things as simple as possible for this MVP.

The API should be simple and declarative. Perhaps something like?:

```js
<Editor theme={theme} components={components} />
```

In order to work with Slate we need to handle a handful of things that are required to integrate into our implementation of the editor.

### Component controls

We can leverage something like `propTypes` or [`propertyControls`](https://blog.framer.com/property-controls-in-framer-x-bdd550b5010c) which determine the props that can be changed and what their dials/knobs might look like.

Depending on the amount of controls and properties allowed, we can open up the inline editor like links/images.

![image](https://user-images.githubusercontent.com/1424573/57143002-1ad3bd80-6d7b-11e9-92b6-1a9d3c20150c.png)

If there are _many_ properties that are editable we can expose a sidebar like the theme editor.

If a component accepts children it can then be edited within as though it's the editor itself.

![image](https://user-images.githubusercontent.com/1424573/57142726-85d0c480-6d7a-11e9-8fad-a16323cf6168.png)

The output MDX will look something like:

```
# Hello, *world*!

<TomatoBox>

## Here's some editing

Inside our custom `TomatoBox` block!

<YouTube videoId="4fHw4GeW3EU" />

</TomatoBox>

And more stuff!
```

In addition to editing the components, the toolbar will need to know what components have been passed and expose them in the toolbar (perhaps in a custom component dropdown or expandable sidebar).

### Commands

We'll need to be able to expose Slate commands so we can insert/toggle components that are in context for when they're clicked. Toggle will be for elements that accept children. Insert for void elements.

### Rendering

In node rendering we will have to check the block type against available components and render them when relevant. If nothing exists, we `return next()` since there might be a plugin that handles the rendering of that node.

### Serialization

This is probably the trickiest piece, but one I think that we have _almost_ there. We have to be able to read in custom components as JSX nodes in the MDX and map them to blocks. This means that the serializer will have to know. [Here's a test showing part of that functionality](https://github.com/blocks/blocks/blob/2f093b968370d8eeed074b6d807176dde77c7349/packages/serializer/test/serializer.test.js).

If it's not a component that it knows about, it sets it as a `jsx` block (for the live editor).

In addition to handling MDX => Slate schema, we need to go the other direction. This part could be a bit tricky since the serializer needs to know, ahead of time, the component block type. One potential way to do that is to add some type of private prop `__BLOCKS_COMPONENT_NODE__: true` with a catch all matcher at the end of the serializer.

If we end up there and the props exist we know to write out as JSX.

### Schema definition

If a component is a void component we will need to specify that in Slate. Currently the schema looks like:

```js
import { Block } from 'slate'

export default {
  document: {
    last: { type: 'paragraph' },
    normalize: (change, { code, node }) => {
      switch (code) {
        case 'last_child_type_invalid': {
          const paragraph = Block.create('paragraph')
          return change.insertNodeByKey(node.key, node.nodes.size, paragraph)
        }
        default: {
        }
      }
    }
  },

  blocks: {
    image: {
      isVoid: true
    },
    hr: {
      isVoid: true
    },
    YouTube: {
      isVoid: true
    },
    Gist: {
      isVoid: true
    }
  }
}
```

We will likely need to have some sort of static property in user components that we can use to extend the config. We could also potentially check in propTypes to see if `children` is accepted as a prop.

---

I think these are all the pieces we need. Am I missing anything @jxnblk?

**Comment by jxnblk** [comment]
This sounds like it covers most of the bases. Here's a few thoughts:

```jsx
// I don't think there's much value in using context here
<Blocks.Provider theme={theme} blockComponents={components}>
  <Blocks.Editor />
</Blocks.Provider>
```

```jsx
// E.g. using props seems more standard/straight-forward
<Editor theme={theme} components={components} />
```

As for the schema and serialization, I think it *could* make sense to treat all blocks as the `jsx` type in the Slate schema, and use an addition `data.type` property or similar to add the root component name, which can be checked against the components provided to the Editor. If we're talking about using static properties to define the shape of the editing forms anyway, we'll need to check that in the runtime. We probably will still need someway to know whether a JSX block accepts children or not though... I think I have enough to put together a proof of concept of this if this doesn't make sense

**Comment by johno** [comment]
Updated the proposed API to `<Editor theme={theme} components={components} />` 👍

**Comment by johno** [comment]
In order to make things simpler we're now opting (as of #17) to only use two custom block types: `jsx` and `jsx-void` which a block extends. The metadata needed for rendering and exposing the form can be added to the node's data field.

```js
{
  type: 'jsx-void',
  data: {
    type: 'YouTube',
    props: {}
  }
}
```

**Comment by johno** [comment]
Addressed by #17 and #19
